Add initial test for RTCPeerConnection ondatachannel

soareschen opened the pull request:
~This is a work in progress.~ (edit: done) I added one basic test for establishing data channel connection with a remote peer and fire a `datachannel` event.

~I also make use of this to test the state change of the peer connection and `RTC*Transport` objects. The idea is that for the `datachannel` event to be fired on a remote peer, it must first establish an SCTP connection and send the data channel negotiation through a `DATA_CHANNEL_OPEN` message. In other words by that time the `RTCDtlsTransport` and `RTCIceTransport` should have all been in connected state.~

~Other than that, since the test connection only have one `RTCDtlsTransport` and one `RTCIceTransport` and we know their final state, we can also test on the peer connection state.~

~There is only one thing that I am not very certain. Currently running on the test on Chrome, the `RTCIceConnectionState` is `connected` instead of `completed`. I am not sure if under TrickleICE, a DTLS connection can be established while the ICE agent is still attempting connection with each candidate pair. The current assumption for the test is that this is allowed, and the assertion allows the final `RTCIceConnectionState` to be either `connected` or `completed`. Would appreciate if anyone familar with this can confirm or correct me if I am wrong.~

(Edit: assumption was wrong and test is removed)

<!-- Reviewable:start -->

<!-- Reviewable:end -->

soareschen committed:
Add helper functions for establishing connection between two pc

soareschen committed:
Add initial test for ondatachannel

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240728836)

# Firefox (nightly) #
Testing web-platform-tests at revision 25dedecc0f460d2dacad77b664042e0f2a6392bf
Using browser at version BuildID 20170607100302; SourceStamp 5801aa478de12a62b2b2982659e787fcc4268d67
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>6 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-createAnswer.html">/webrtc/RTCPeerConnection-createAnswer.html</a></summary>


|                                      Subtest                                      | Results | Messages |
|-----------------------------------------------------------------------------------|---------|----------|
|                                                                                   | OK      |          |
| `createAnswer() with null remoteDescription should reject with InvalidStateError` | PASS    |          |
| `createAnswer() after setting remote description should succeed`                  | PASS    |          |
| `createAnswer() when connection is closed reject with InvalidStateError`          | PASS    |          |
| `createAnswer() when connection is closed in parallel should never resolve`       | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-createOffer.html">/webrtc/RTCPeerConnection-createOffer.html</a></summary>


|                                                               Subtest                                                               | Results |                                                                                Messages                                                                                |
|-------------------------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                                     | OK      |                                                                                                                                                                        |
| `createOffer() with no argument from newly created RTCPeerConnection should succeed`                                                | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."` |
| `createOffer() and then setLocalDescription() should succeed`                                                                       | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                        |
| `createOffer() after connection is closed should reject with InvalidStateError`                                                     | PASS    |                                                                                                                                                                        |
| `createOffer() when connection is closed halfway should never resolve`                                                              | PASS    |                                                                                                                                                                        |
| `When media stream is added when createOffer() is running in parallel, the result offer should contain the new media stream`        | FAIL    | `pc.addTransceiver is not a function`                                                                                                                                  |
| `createOffer() with offerToReceiveAudio should add audio line to all subsequent created offers`                                     | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."` |
| `createOffer() with offerToReceiveVideo should add video line to all subsequent created offers`                                     | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."` |
| `createOffer() with offerToReceiveAudio:true then offerToReceiveVideo:true should have result offer with both audio and video line` | FAIL    | `assert_equals: Expect audio line to remain in created offer expected 1 but got 0`                                                                                     |
| `Test onsignalingstatechange event for createOffer() and then setLocalDescription() should succeed`                                 | PASS    |                                                                                                                                                                        |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-ondatachannel.html">/webrtc/RTCPeerConnection-ondatachannel.html</a></summary>


|                                        Subtest                                        | Results |             Messages            |
|---------------------------------------------------------------------------------------|---------|---------------------------------|
|                                                                                       | OK      |                                 |
| `datachannel event should fire when new data channel is announced to the remote peer` | FAIL    | `RTCDataChannel is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-setLocalDescription.html">/webrtc/RTCPeerConnection-setLocalDescription.html</a></summary>


|                                                                          Subtest                                                                          | Results |                                                                                                                              Messages                                                                                                                             |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------|---------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                                                           | OK      |                                                                                                                                                                                                                                                                   |
| `setLocalDescription with type offer and null sdp should use lastOffer generated from createOffer`                                                        | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."`                                                                                            |
| `setLocalDescription() with offer not created by own createOffer() should reject with InvalidModificationError`                                           | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidStateError: Calling SetLocal without first calling CreateOffer/Answer is not supported." that is not a DOMException InvalidModificationError: property "code" is equal to 11, expected 13` |
| `Set created offer other than last offer should reject with InvalidModificationError`                                                                     | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidSessionDescriptionError: Changing the media-type of m-sections is not allowed" that is not a DOMException InvalidModificationError: property "code" is equal to 0, expected 13`            |
| `Creating and setting offer multiple times should succeed`                                                                                                | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidStateError: Cannot create offer in state have-local-offer"`                                                                                                                                         |
| `setLocalDescription(offer) should never resolve if connection is closed in parallel`                                                                     | FAIL    | `assert_unreached: Pending promise should never be resolved. Instead it is rejected with: InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel. Reached unreachable code`                                 |
| `setLocalDescription() with valid answer should succeed`                                                                                                  | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                                                                                                                   |
| `setLocalDescription() with type answer and null sdp should use lastAnswer generated from createAnswer`                                                   | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidParameterError: Empty or null SDP provided to setLocalDescription"`                                                                                                                                 |
| `setLocalDescription() with answer not created by own createAnswer() should reject with InvalidModificationError`                                         | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidStateError: Calling SetLocal without first calling CreateOffer/Answer is not supported." that is not a DOMException InvalidModificationError: property "code" is equal to 11, expected 13` |
| `Calling createOffer() and setLocalDescription() again after one round of local-offer/remote-answer should succeed`                                       | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                                                                                                                   |
| `Switching role from answerer to offerer after going back to stable state should succeed`                                                                 | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                                                                                                                   |
| `Calling setLocalDescription(answer) from stable state should reject with InvalidStateError`                                                              | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."`                                                                                            |
| `Calling setLocalDescription(answer) from have-local-offer state should reject with InvalidStateError`                                                    | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."`                                                                                            |
| `Test onsignalingstatechange event for Creating and setting offer multiple times should succeed`                                                          | PASS    |                                                                                                                                                                                                                                                                   |
| `Test onsignalingstatechange event for setLocalDescription() with valid answer should succeed`                                                            | PASS    |                                                                                                                                                                                                                                                                   |
| `Test onsignalingstatechange event for Calling createOffer() and setLocalDescription() again after one round of local-offer/remote-answer should succeed` | PASS    |                                                                                                                                                                                                                                                                   |
| `Test onsignalingstatechange event for Switching role from answerer to offerer after going back to stable state should succeed`                           | PASS    |                                                                                                                                                                                                                                                                   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-setRemoteDescription.html">/webrtc/RTCPeerConnection-setRemoteDescription.html</a></summary>


|                                                                 Subtest                                                                  | Results |                                                                                                              Messages                                                                                                             |
|------------------------------------------------------------------------------------------------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                                          | TIMEOUT |                                                                                                                                                                                                                                   |
| `setRemoteDescription with valid offer should succeed`                                                                                   | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                                                                                   |
| `Setting remote description multiple times with different offer should succeed`                                                          | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidStateError: Cannot set remote offer or answer in current state have-remote-offer"`                                                                                  |
| `setRemoteDescription(offer) should never resolve if connection is closed in parallel`                                                   | FAIL    | `assert_unreached: Pending promise should never be resolved. Instead it is rejected with: InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel. Reached unreachable code` |
| `setRemoteDescription(offer) with invalid SDP should reject with RTCError`                                                               | FAIL    | `assert_equals: Expect error detail field to set to sdp-syntax-error expected (string) "sdp-syntax-error" but got (undefined) undefined`                                                                                          |
| `setRemoteDescription with invalid type and invalid SDP should reject with TypeError`                                                    | PASS    |                                                                                                                                                                                                                                   |
| `setRemoteDescription() with invalid SDP and stable state should reject with InvalidStateError`                                          | PASS    |                                                                                                                                                                                                                                   |
| `setRemoteDescription() with valid state and answer should succeed`                                                                      | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                                                                                   |
| `Calling setRemoteDescription() again after one round of remote-offer/local-answer should succeed`                                       | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidSessionDescriptionError: Remote description changes the media type of m-line 0"`                                                                                    |
| `Switching role from offerer to answerer after going back to stable state should succeed`                                                | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                                                                                   |
| `Calling setRemoteDescription(answer) from stable state should reject with InvalidStateError`                                            | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."`                                                            |
| `Calling setRemoteDescription(offer) from have-local-offer state should reject with InvalidStateError`                                   | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."`                                                            |
| `Calling setRemoteDescription(answer) from have-remote-offer state should reject with InvalidStateError`                                 | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."`                                                            |
| `setRemoteDescription should trigger ontrack event when the MSID of the stream is is parsed.`                                            | FAIL    | `assert_own_property: Expect pc to have ontrack event handler attribute expected property "ontrack" missing`                                                                                                                      |
| `Test onsignalingstatechange event for setRemoteDescription with valid offer should succeed`                                             | PASS    |                                                                                                                                                                                                                                   |
| `Test onsignalingstatechange event for Setting remote description multiple times with different offer should succeed`                    | PASS    |                                                                                                                                                                                                                                   |
| `Test onsignalingstatechange event for setRemoteDescription() with valid state and answer should succeed`                                | PASS    |                                                                                                                                                                                                                                   |
| `Test onsignalingstatechange event for Calling setRemoteDescription() again after one round of remote-offer/local-answer should succeed` | TIMEOUT | `Test timed out`                                                                                                                                                                                                                  |
| `Test onsignalingstatechange event for Switching role from offerer to answerer after going back to stable state should succeed`          | PASS    |                                                                                                                                                                                                                                   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCSctpTransport-constructor.html">/webrtc/RTCSctpTransport-constructor.html</a></summary>


|                                       Subtest                                        | Results |                                Messages                               |
|--------------------------------------------------------------------------------------|---------|-----------------------------------------------------------------------|
|                                                                                      | OK      |                                                                       |
| `setRemoteDescription() with answer containing data media should initialize pc.sctp` | FAIL    | `assert_equals: expected (object) null but got (undefined) undefined` |
| `setLocalDescription() with answer containing data media should initialize pc.sctp`  | FAIL    | `assert_equals: expected (object) null but got (undefined) undefined` |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240728837)

# Sauce (safari) #
Testing web-platform-tests at revision 25dedecc0f460d2dacad77b664042e0f2a6392bf
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>6 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-createAnswer.html">/webrtc/RTCPeerConnection-createAnswer.html</a></summary>


|                                      Subtest                                      | Results |                 Messages                 |
|-----------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                   | OK      |                                          |
| `createAnswer() with null remoteDescription should reject with InvalidStateError` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createAnswer() after setting remote description should succeed`                  | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createAnswer() when connection is closed reject with InvalidStateError`          | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createAnswer() when connection is closed in parallel should never resolve`       | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-createOffer.html">/webrtc/RTCPeerConnection-createOffer.html</a></summary>


|                                                               Subtest                                                               | Results |                 Messages                 |
|-------------------------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                                                     | OK      |                                          |
| `createOffer() with no argument from newly created RTCPeerConnection should succeed`                                                | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createOffer() and then setLocalDescription() should succeed`                                                                       | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createOffer() after connection is closed should reject with InvalidStateError`                                                     | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createOffer() when connection is closed halfway should never resolve`                                                              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `When media stream is added when createOffer() is running in parallel, the result offer should contain the new media stream`        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createOffer() with offerToReceiveAudio should add audio line to all subsequent created offers`                                     | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createOffer() with offerToReceiveVideo should add video line to all subsequent created offers`                                     | FAIL    | `Can't find variable: RTCPeerConnection` |
| `createOffer() with offerToReceiveAudio:true then offerToReceiveVideo:true should have result offer with both audio and video line` | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-ondatachannel.html">/webrtc/RTCPeerConnection-ondatachannel.html</a></summary>


|                                        Subtest                                        | Results |                 Messages                 |
|---------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                       | OK      |                                          |
| `datachannel event should fire when new data channel is announced to the remote peer` | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-setLocalDescription.html">/webrtc/RTCPeerConnection-setLocalDescription.html</a></summary>


|                                                       Subtest                                                       | Results |                 Messages                 |
|---------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                                     | OK      |                                          |
| `setLocalDescription with type offer and null sdp should use lastOffer generated from createOffer`                  | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription() with offer not created by own createOffer() should reject with InvalidModificationError`     | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Set created offer other than last offer should reject with InvalidModificationError`                               | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Creating and setting offer multiple times should succeed`                                                          | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription(offer) should never resolve if connection is closed in parallel`                               | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription() with valid answer should succeed`                                                            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription() with type answer and null sdp should use lastAnswer generated from createAnswer`             | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription() with answer not created by own createAnswer() should reject with InvalidModificationError`   | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling createOffer() and setLocalDescription() again after one round of local-offer/remote-answer should succeed` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Switching role from answerer to offerer after going back to stable state should succeed`                           | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setLocalDescription(answer) from stable state should reject with InvalidStateError`                        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setLocalDescription(answer) from have-local-offer state should reject with InvalidStateError`              | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCPeerConnection-setRemoteDescription.html">/webrtc/RTCPeerConnection-setRemoteDescription.html</a></summary>


|                                                 Subtest                                                  | Results |                 Messages                 |
|----------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                          | OK      |                                          |
| `setRemoteDescription with valid offer should succeed`                                                   | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Setting remote description multiple times with different offer should succeed`                          | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setRemoteDescription(offer) should never resolve if connection is closed in parallel`                   | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setRemoteDescription(offer) with invalid SDP should reject with RTCError`                               | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setRemoteDescription with invalid type and invalid SDP should reject with TypeError`                    | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setRemoteDescription() with invalid SDP and stable state should reject with InvalidStateError`          | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setRemoteDescription() with valid state and answer should succeed`                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setRemoteDescription() again after one round of remote-offer/local-answer should succeed`       | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Switching role from offerer to answerer after going back to stable state should succeed`                | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setRemoteDescription(answer) from stable state should reject with InvalidStateError`            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setRemoteDescription(offer) from have-local-offer state should reject with InvalidStateError`   | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setRemoteDescription(answer) from have-remote-offer state should reject with InvalidStateError` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setRemoteDescription should trigger ontrack event when the MSID of the stream is is parsed.`            | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6186/webrtc/RTCSctpTransport-constructor.html">/webrtc/RTCSctpTransport-constructor.html</a></summary>


|                                       Subtest                                        | Results |                 Messages                 |
|--------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                      | OK      |                                          |
| `setRemoteDescription() with answer containing data media should initialize pc.sctp` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription() with answer containing data media should initialize pc.sctp`  | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240728843)

# Chrome (unstable) #
Testing web-platform-tests at revision 25dedecc0f460d2dacad77b664042e0f2a6392bf

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240728846)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 25dedecc0f460d2dacad77b664042e0f2a6392bf

rwaldron commented on webrtc/RTCPeerConnection-helper.js:
```diff
@@ -210,3 +210,37 @@ function test_never_resolve(testFunc, testName) {
+// Helper function to exchange ice candidates between
+// two local peer connections
+function exchangeIceCandidates(pc1, pc2) {
+  function doExchange(localPc, remotePc) {
```
Move `doExchange` outside of `exchangeIceCandidates` (there are no closures)

soareschen replied:
I know. But without proper module system I don't really want to expose the private helper function globally. Putting it as an inner function helps indicate that it is private without going through too much hacks.

rwaldron replied:
Ok, that's a fair argument ;P

rwaldron commented on webrtc/RTCPeerConnection-ondatachannel.html:
```diff
@@ -0,0 +1,225 @@
+
+  // The following helper functions are called from RTCPeerConnection-helper.js:
+  // exchangeIceCandidates
+  // doSignalingHandshake
```
👍

rwaldron reviewed (changes requested):
Small fix. I know you're not done yet, but I've begun updating the coverage spec as well: https://rwaldron.github.io/webrtc-pc/#rtcdatachannel (I'll keep tracking as you progress)

soareschen committed:
Remove test on connection states as the assumptions are wrong

soareschen committed:
Add more test for ondatachannel

soareschen commented:
I have rebased and added a few more tests.

I also removed the tests on the connection states. Turns out my assumption was wrong and the underlying transports can be of any state regardless of whether the upper layer transport is connected. Will figure other way to test the peer connection states.

rwaldron commented:
Great work!
